galley/pkg/server/process/host.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package process
+
+import (
+	"errors"
+	"sync"
+)
+
+// Host is a process host that takes care of start/stop of sub-components.
+type Host struct {
+	mu         sync.Mutex
+	started    bool
+	components []Component
+}
+
+// Add a new component to this host. It panics if the host is already started.
+func (h *Host) Add(c Component) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.started {
+		panic("Host.Add: host is already started")
+	}
+	h.components = append(h.components, c)
+}
+
+// Start the process.Host. If any of the components starts to fail, then an error is returned. If there is an error
+// then all successfully started components will be stopped before control is returned.
+func (h *Host) Start() error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	if h.started {
+		return errors.New("process.Host: already started")
+	}
+
+	for i := 0; i < len(h.components); i++ {
+		c := h.components[i]
+		if err := c.Start(); err != nil {
+			// Component startup failed. Stop already started components.
+			for j := i - 1; j >= 0; j-- {
+				h.components[j].Stop()
+			}
+			return err
+		}
+	}
+	h.started = true
+
+	return nil
+}
+
+// Stop the Host. All components will be stopped, before the control is returned.
```

Comment:
Do we need to guard against the possibility that a component may take unexpectedly long to stop (or start)?

Reply:
No, not really. All of these components are written by us, so trying to protect against long start/stop time at this level amounts to protecting against our own code bugs. It is better to fix whatever is causing those long timings.